File: anvil/backend/store.py
"""Durable, file-based data store — everything Anvil accumulates at runtime.

  data/
    builds/<build_id>/
      build.json        manifest: requirement, model, coverage, timing, summary
      events.jsonl      every loop event that streamed to the UI
      api_calls.jsonl   every Anthropic API call (full request + response + usage)
    kb_learned.json     web-discovered components, deduped — the KB grows here
    builds_index.jsonl  one line per build, for fast listing

No database. Append-only files a human can read and a judge can watch grow.
Each /api/run is treated as one hardware BUILD with its own folder.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path

from .schema import Component, all_components

_DEFAULT_DATA_DIR = Path(__file__).resolve().parent.parent / "data"


def _root() -> Path:
    return Path(os.environ.get("ANVIL_DATA_DIR", str(_DEFAULT_DATA_DIR)))


def _builds_dir() -> Path:
    return _root() / "builds"


def _learned_path() -> Path:
    return _root() / "kb_learned.json"


def _index_path() -> Path:
    return _root() / "builds_index.jsonl"


def _ensure():
    _builds_dir().mkdir(parents=True, exist_ok=True)
    if not _learned_path().exists():
        _learned_path().write_text("[]", encoding="utf-8")


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def new_build_id() -> str:
    stamp = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
    return f"build-{stamp}-{uuid.uuid4().hex[:4]}"


# --------------------------------------------------------------------------- #
# per-build recorder
# --------------------------------------------------------------------------- #

class BuildRecorder:
    """Records one hardware build: its event stream and every API call it made."""

    def __init__(self, build_id: str, requirement: dict, model: str):
        _ensure()
        self.id = build_id
        self.requirement = requirement
        self.model = model
        self.started = _now_iso()
        self.t0 = time.time()
        self.dir = _builds_dir() / build_id
        self.dir.mkdir(parents=True, exist_ok=True)
        self._events = self.dir / "events.jsonl"
        self._api = self.dir / "api_calls.jsonl"
        self.n_events = 0
        self.n_api_calls = 0

    def log_event(self, ev: dict):
        self.n_events += 1
        with self._events.open("a", encoding="utf-8") as f:
            f.write(json.dumps(ev) + "\n")

    def log_api_call(self, rec: dict):
        self.n_api_calls += 1
        rec = {"ts": _now_iso(), "build_id": self.id, **rec}
        with self._api.open("a", encoding="utf-8") as f:
            f.write(json.dumps(rec, default=str) + "\n")

    def finalize(self, done_event: dict | None):
        spec = (done_event or {}).get("spec")
        manifest = {
            "build_id": self.id,
            "name": self.requirement.get("name", "untitled"),
            "model": self.model,
            "requirement": self.requirement,
            "started": self.started,
            "finished": _now_iso(),
            "elapsed_s": round(time.time() - self.t0, 2),
            "coverage": (done_event or {}).get("coverage", 0.0),
            "soft_score": (done_event or {}).get("soft_score", 0.0),
            "iterations": (done_event or {}).get("iterations", 0),
            "all_hard_pass": bool((done_event or {}).get("all_hard_pass")),
            "new_rules": (done_event or {}).get("new_rules", []),
            "api_calls": self.n_api_calls,
            "events": self.n_events,
            "final_bom": (spec or {}).get("bom"),
        }
        (self.dir / "build.json").write_text(json.dumps(manifest, indent=2), encoding="utf-8")
        with _index_path().open("a", encoding="utf-8") as f:
            f.write(json.dumps({k: manifest[k] for k in (
                "build_id", "name", "model", "started", "elapsed_s",
                "coverage", "iterations", "all_hard_pass", "api_calls")}) + "\n")
        return manifest


# --------------------------------------------------------------------------- #
# build listing / retrieval
# --------------------------------------------------------------------------- #

def list_builds(limit: int = 50) -> list[dict]:
    _ensure()
    if not _index_path().exists():
        return []
    rows = [json.loads(l) for l in _index_path().read_text(encoding="utf-8").splitlines() if l.strip()]
    return list(reversed(rows))[:limit]


def get_build(build_id: str) -> dict | None:
    d = _builds_dir() / build_id
    man = d / "build.json"
    if not man.exists():
        return None
    out = json.loads(man.read_text(encoding="utf-8"))
    ev = d / "events.jsonl"
    out["event_log"] = [json.loads(l) for l in ev.read_text(encoding="utf-8").splitlines() if l.strip()] if ev.exists() else []
    return out


def get_build_api_calls(build_id: str) -> list[dict]:
    f = _builds_dir() / build_id / "api_calls.jsonl"
    if not f.exists():
        return []
    return [json.loads(l) for l in f.read_text(encoding="utf-8").splitlines() if l.strip()]


# --------------------------------------------------------------------------- #
# learned KB — web-discovered specs accumulate here
# --------------------------------------------------------------------------- #

def load_learned() -> list[Component]:
    _ensure()
    try:
        raw = json.loads(_learned_path().read_text(encoding="utf-8"))
    except (json.JSONDecodeError, FileNotFoundError):
        return []
    return [Component.from_dict(r) for r in raw]


def merge_learned_into(kb: dict[str, list[Component]]) -> int:
    """Merge the learned library into an in-memory KB dict (dedup by id). Returns count merged."""
    have = {p.id for parts in kb.values() for p in parts}
    n = 0
    for c in load_learned():
        if c.id not in have:
            kb.setdefault(c.subsystem, []).append(c)
            have.add(c.id)
            n += 1
    return n


def harvest_web_parts(bom: dict, kb: dict[str, list[Component]] | None = None) -> list[dict]:
    """Persist any source='web' components from a finished BOM into the learned
    library (dedup by id) and, if a live KB is given, merge them in so the next
    run can reuse them. Returns the newly stored components as dicts."""
    _ensure()
    existing = {c.id for c in load_learned()}
    in_kb = {p.id for parts in (kb or {}).values() for p in parts} if kb else set()
    new: list[Component] = []
    for comp in all_components(bom):
        if comp.source == "web" and comp.id not in existing:
            existing.add(comp.id)
            new.append(comp)
    if not new:
        return []
    # rewrite the learned file with the union
    current = load_learned() + new
    _learned_path().write_text(json.dumps([c.to_dict() for c in current], indent=2), encoding="utf-8")
    # merge into the live KB
    if kb is not None:
        for c in new:
            if c.id not in in_kb:
                kb.setdefault(c.subsystem, []).append(c)
    return [c.to_dict() for c in new]


def learned_count() -> int:
    return len(load_learned())

File: web/lib/store.ts
// Durable, file-based data store — everything Anvil accumulates at runtime
// (faithful port of anvil/backend/store.py).
//
//   data/
//     builds/<build_id>/
//       build.json        manifest: requirement, model, coverage, timing, summary
//       events.jsonl      every loop event that streamed to the UI
//       api_calls.jsonl   every Anthropic API call (full request + response + usage)
//     kb_learned.json     web-discovered components, deduped — the KB grows here
//     builds_index.jsonl  one line per build, for fast listing
//
// No database. Append-only files a human can read and a judge can watch grow.
// Each /api/run is treated as one hardware BUILD with its own folder.
//
// Vercel reality: the bundled `web/data` tree is read-only. Every WRITE targets
// os.tmpdir()/anvil and is wrapped in try/catch so a read-only FS NEVER throws
// on a request. Every READ merges the bundled seed under web/data with whatever
// has accumulated under tmp (tmp wins on id collisions).

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { allComponents, type BOM, type Component } from "@/lib/types";

// --------------------------------------------------------------------------- //
// roots: bundled seed (read-only) + tmp (writable runtime)
// --------------------------------------------------------------------------- //

function seedRoot(): string {
  return path.join(process.cwd(), "data");
}

function tmpRoot(): string {
  return path.join(os.tmpdir(), "anvil");
}

function seedBuildsDir(): string {
  return path.join(seedRoot(), "builds");
}

function tmpBuildsDir(): string {
  return path.join(tmpRoot(), "builds");
}

function seedLearnedPath(): string {
  return path.join(seedRoot(), "kb_learned.json");
}

function tmpLearnedPath(): string {
  return path.join(tmpRoot(), "kb_learned.json");
}

function seedIndexPath(): string {
  return path.join(seedRoot(), "builds_index.jsonl");
}

function tmpIndexPath(): string {
  return path.join(tmpRoot(), "builds_index.jsonl");
}

// --------------------------------------------------------------------------- //
// fs helpers — all writes best-effort, never throw on read-only FS
// --------------------------------------------------------------------------- //

function readTextSafe(p: string): string | null {
  try {
    return fs.readFileSync(p, "utf-8");
  } catch {
    return null;
  }
}

function ensureDir(dir: string): boolean {
  try {
    fs.mkdirSync(dir, { recursive: true });
    return true;
  } catch {
    return false;
  }
}

function appendLineSafe(p: string, line: string): void {
  if (!ensureDir(path.dirname(p))) return;
  try {
    fs.appendFileSync(p, line + "\n", "utf-8");
  } catch {
    // read-only FS — swallow.
  }
}

function writeTextSafe(p: string, text: string): void {
  if (!ensureDir(path.dirname(p))) return;
  try {
    fs.writeFileSync(p, text, "utf-8");
  } catch {
    // read-only FS — swallow.
  }
}

function parseJsonl<T = Record<string, unknown>>(text: string | null): T[] {
  if (!text) return [];
  const out: T[] = [];
  for (const ln of text.split(/\r?\n/)) {
    const s = ln.trim();
    if (!s) continue;
    try {
      out.push(JSON.parse(s) as T);
    } catch {
      // skip malformed line
    }
  }
  return out;
}

function nowIso(): string {
  // ISO-8601 to seconds precision, mirroring Python isoformat(timespec="seconds").
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function pad(n: number, width = 2): string {
  return String(n).padStart(width, "0");
}

/** "build-YYYYMMDD-HHMMSS-xxxx" — same shape as store.py new_build_id(). */
export function newBuildId(): string {
  const d = new Date();
  const stamp =
    `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1)}${pad(d.getUTCDate())}` +
    `-${pad(d.getUTCHours())}${pad(d.getUTCMinutes())}${pad(d.getUTCSeconds())}`;
  const suffix = Math.random().toString(16).slice(2, 6).padEnd(4, "0");
  return `build-${stamp}-${suffix}`;
}

// --------------------------------------------------------------------------- //
// per-build recorder — writes to tmp only, best-effort
// --------------------------------------------------------------------------- //

export interface BuildManifest {
  build_id: string;
  name: string;
  model: string;
  requirement: Record<string, unknown>;
  started: string;
  finished: string;
  elapsed_s: number;
  coverage: number;
  soft_score: number;
  iterations: number;
  all_hard_pass: boolean;
  new_rules: unknown[];
  api_calls: number;
  events: number;
  final_bom: BOM | null;
}

/** Records one hardware build: its event stream and every API call it made. */
export class BuildRecorder {
  readonly id: string;
  readonly requirement: Record<string, unknown>;
  readonly model: string;
  readonly started: string;
  private readonly t0: number;
  private readonly dir: string;
  private readonly eventsPath: string;
  private readonly apiPath: string;
  nEvents = 0;
  nApiCalls = 0;

  constructor(buildId: string, requirement: Record<string, unknown>, model: string) {
    this.id = buildId;
    this.requirement = requirement;
    this.model = model;
    this.started = nowIso();
    this.t0 = Date.now();
    this.dir = path.join(tmpBuildsDir(), buildId);
    this.eventsPath = path.join(this.dir, "events.jsonl");
    this.apiPath = path.join(this.dir, "api_calls.jsonl");
    ensureDir(this.dir);
  }

  logEvent(ev: Record<string, unknown>): void {
    this.nEvents += 1;
    appendLineSafe(this.eventsPath, JSON.stringify(ev));
  }

  logApiCall(rec: Record<string, unknown>): void {
    this.nApiCalls += 1;
    const withMeta = { ts: nowIso(), build_id: this.id, ...rec };
    appendLineSafe(this.apiPath, JSON.stringify(withMeta));
  }

  finalize(doneEvent: Record<string, unknown> | null): BuildManifest {
    const done = doneEvent ?? {};
    const spec = (done.spec as { bom?: BOM } | undefined) ?? undefined;
    const manifest: BuildManifest = {
      build_id: this.id,
      name: (this.requirement.name as string) ?? "untitled",
      model: this.model,
      requirement: this.requirement,
      started: this.started,
      finished: nowIso(),
      elapsed_s: Math.round((Date.now() - this.t0) / 10) / 100,
      coverage: (done.coverage as number) ?? 0.0,
      soft_score: (done.soft_score as number) ?? 0.0,
      iterations: (done.iterations as number) ?? 0,
      all_hard_pass: Boolean(done.all_hard_pass),
      new_rules: (done.new_rules as unknown[]) ?? [],
      api_calls: this.nApiCalls,
      events: this.nEvents,
      final_bom: spec?.bom ?? null,
    };
    writeTextSafe(
      path.join(this.dir, "build.json"),
      JSON.stringify(manifest, null, 2),
    );
    const indexRow = {
      build_id: manifest.build_id,
      name: manifest.name,
      model: manifest.model,
      started: manifest.started,
      elapsed_s: manifest.elapsed_s,
      coverage: manifest.coverage,
      iterations: manifest.iterations,
      all_hard_pass: manifest.all_hard_pass,
      api_calls: manifest.api_calls,
    };
    appendLineSafe(tmpIndexPath(), JSON.stringify(indexRow));
    return manifest;
  }
}

// --------------------------------------------------------------------------- //
// build listing / retrieval — merges seed + tmp
// --------------------------------------------------------------------------- //

/** Most-recent-first list of build index rows, capped at `limit`. Merges the
 *  bundled seed index with tmp-accumulated builds (tmp wins on build_id). */
export function listBuilds(limit = 50): Array<Record<string, unknown>> {
  const seed = parseJsonl(readTextSafe(seedIndexPath()));
  const tmp = parseJsonl(readTextSafe(tmpIndexPath()));
  // tmp rows are newer; let them override seed rows with the same build_id.
  const byId = new Map<string, Record<string, unknown>>();
  const order: string[] = [];
  for (const row of [...seed, ...tmp]) {
    const id = String((row as Record<string, unknown>).build_id ?? "");
    if (!id) continue;
    if (!byId.has(id)) order.push(id);
    byId.set(id, row as Record<string, unknown>);
  }
  const merged = order.map((id) => byId.get(id)!);
  return merged.reverse().slice(0, limit);
}

function buildDirFor(buildId: string): string | null {
  // tmp wins (a build re-run there is newer than the bundled copy).
  const tmpD = path.join(tmpBuildsDir(), buildId);
  if (fs.existsSync(path.join(tmpD, "build.json"))) return tmpD;
  const seedD = path.join(seedBuildsDir(), buildId);
  if (fs.existsSync(path.join(seedD, "build.json"))) return seedD;
  return null;
}

/** Full build manifest with its `event_log` inlined, or null if unknown. */
export function getBuild(buildId: string): Record<string, unknown> | null {
  const dir = buildDirFor(buildId);
  if (!dir) return null;
  const manText = readTextSafe(path.join(dir, "build.json"));
  if (!manText) return null;
  let out: Record<string, unknown>;
  try {
    out = JSON.parse(manText) as Record<string, unknown>;
  } catch {
    return null;
  }
  out.event_log = parseJsonl(readTextSafe(path.join(dir, "events.jsonl")));
  return out;
}

/** Every recorded Anthropic API call for a build (empty list if none). */
export function getBuildApiCalls(buildId: string): Array<Record<string, unknown>> {
  const dir = buildDirFor(buildId);
  if (!dir) return [];
  return parseJsonl(readTextSafe(path.join(dir, "api_calls.jsonl")));
}

// --------------------------------------------------------------------------- //
// learned KB — web-discovered specs accumulate here (seed + tmp)
// --------------------------------------------------------------------------- //

function parseLearned(text: string | null): Component[] {
  if (!text) return [];
  try {
    const raw = JSON.parse(text);
    return Array.isArray(raw) ? (raw as Component[]) : [];
  } catch {
    return [];
  }
}

/** All learned components: bundled seed first, then tmp-accumulated, deduped by
 *  id (first occurrence wins, so seed is authoritative for shared ids). */
export function loadLearned(): Component[] {
  const seen = new Set<string>();
  const out: Component[] = [];
  for (const c of [
    ...parseLearned(readTextSafe(seedLearnedPath())),
    ...parseLearned(readTextSafe(tmpLearnedPath())),
  ]) {
    if (!c || !c.id || seen.has(c.id)) continue;
    seen.add(c.id);
    out.push(c);
  }
  return out;
}

/** Merge the learned library into an in-memory KB dict (dedup by id). Returns
 *  the count merged. Mirrors store.py merge_learned_into. */
export function mergeLearnedInto(kb: Record<string, Component[]>): number {
  const have = new Set<string>();
  for (const parts of Object.values(kb)) for (const p of parts) have.add(p.id);
  let n = 0;
  for (const c of loadLearned()) {
    if (!have.has(c.id)) {
      (kb[c.subsystem] ??= []).push(c);
      have.add(c.id);
      n += 1;
    }
  }
  return n;
}

/** Persist any source='web' components from a finished BOM into the learned
 *  library (dedup by id) and, if a live KB is given, merge them in so the next
 *  run can reuse them. Returns the newly stored components. Best-effort write. */
export function harvestWebParts(
  bom: BOM,
  kb: Record<string, Component[]> | null = null,
): Component[] {
  const existing = new Set(loadLearned().map((c) => c.id));
  const inKb = new Set<string>();
  if (kb) for (const parts of Object.values(kb)) for (const p of parts) inKb.add(p.id);

  const fresh: Component[] = [];
  for (const comp of allComponents(bom)) {
    if (comp.source === "web" && comp.id && !existing.has(comp.id)) {
      existing.add(comp.id);
      fresh.push(comp);
    }
  }
  if (fresh.length === 0) return [];

  // rewrite the tmp learned file with the union (seed + tmp + new).
  const union = [...loadLearned(), ...fresh];
  writeTextSafe(tmpLearnedPath(), JSON.stringify(union, null, 2));

  // merge into the live KB
  if (kb) {
    for (const c of fresh) {
      if (!inKb.has(c.id)) (kb[c.subsystem] ??= []).push(c);
    }
  }
  return fresh;
}

/** Total count of learned (web-harvested) components across seed + tmp. */
export function learnedCount(): number {
  return loadLearned().length;
}
